lib/srv/db/watcher.go

```diff
@@ -0,0 +1,119 @@
+	log := s.log.WithField("selector", s.cfg.Selectors)
+	if len(s.cfg.Selectors) == 0 {
+		log.Debug("Not initializing database resource watcher.")
+		return nil, nil
+	}
+	log.Debug("Initializing database resource watcher.")
+	watcher, err := services.NewDatabaseWatcher(ctx, services.DatabaseWatcherConfig{
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component: teleport.ComponentDatabase,
+			Log:       log,
+			Client:    s.cfg.AccessPoint,
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	go func() {
+		defer watcher.Close()
+		for {
+			select {
+			case databases := <-watcher.DatabasesC:
+				if err := s.reconcileResources(ctx, databases); err != nil {
+					log.WithError(err).Errorf("Failed to reconcile %v.", databases)
+				} else if s.cfg.OnReconcile != nil {
+					s.cfg.OnReconcile(s.getDatabases())
+				}
+			case <-ctx.Done():
+				log.Debug("Database resource watcher done.")
+				return
+			}
+		}
+	}()
+	return watcher, nil
+}
+
+// reconcileResources reconciles the database resources this server is currently
+// proxying with the provided up-to-date list of cluster databases.
+func (s *Server) reconcileResources(ctx context.Context, newResources types.Databases) error {
+	s.log.Debugf("Reconciling with %v database resources.", len(newResources))
+	// First remove databases that aren't present in the resource list anymore.
+	for _, current := range s.getDatabases() {
+		// Skip databases from static configuration. For backwards compatibility
+		// also consider empty "origin" value.
+		if current.Origin() == types.OriginConfigFile || current.Origin() == "" {
+			continue
+		}
+		if new := newResources.Find(current.GetName()); new == nil {
+			s.log.Infof("%v removed, unregistering.", current)
+			if err := s.unregisterDatabase(ctx, current.GetName()); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+	// Then add new databases if there are any or refresh those that were updated.
+	for _, new := range newResources {
+		if db := s.getDatabases().Find(new.GetName()); db == nil {
+			// Database with this name isn't registered yet, see if it matches.
+			if services.MatchDatabase(s.cfg.Selectors, new) {
+				s.log.Infof("%v matches, registering.", new)
+				if err := s.registerDatabase(ctx, new); err != nil {
+					return trace.Wrap(err)
+				}
+			} else {
+				s.log.Debugf("%v doesn't match, not registering.", new)
+			}
+		} else if db.Origin() == types.OriginConfigFile || db.Origin() == "" {
+			// Do not overwrite databases from static configuration.
+			s.log.Infof("%v is part of static configuration, not registering %v.", db, new)
+			continue
+		} else if new.GetResourceID() != db.GetResourceID() {
+			// If labels were updated, the database may no longer match.
+			if services.MatchDatabase(s.cfg.Selectors, new) {
+				s.log.Infof("%v updated, re-registering.", new)
+				if err := s.registerDatabase(ctx, new); err != nil {
```

Comment:
Does calling registerDatabase on already registered database work and resources are cleaned  properly ?

Reply:
Good call, re-registering worked but it wasn't cleaned up properly. Updated to stop heartbeat and label updates when re-registering.